In [1]:
import sys
sys.path.append('./lib/')

with open('./lib/common_imports.py') as f:
    exec(f.read())

from hypothesis_test import pttest

# Hypothesis testing

To make decisions using data, we need to characterize the kinds of conclusions we can make. Classical hypothesis testing is concerned with deciding between two decisions (things get much harder if there’s more than two). The first, a null hypothesis is specified that represents the status quo. This hypothesis is usually labeled, $H_{0}$. This is what we assume by default. The alternative or
research hypothesis is what we require evidence to conclude. This hypothesis is usually labeled, $H_{a}$ or sometimes $H_{1}$ (or some other number other than $0$). So to reiterate, the null hypothesis is assumed true and statistical evidence is required to reject it in favor of a research or alternative hypothesis.

## Example

A respiratory disturbance index (RDI) of more than $30$ events / hour, say, is considered evidence of severe sleep disordered breathing (SDB). Suppose that in a sample of $100$ overweight subjects with
other risk factors for sleep disordered breathing at a sleep clinic, the mean RDI was $32$ events / hour with a standard deviation of $10$ events / hour. We might want to test the hypothesis that

$$
H_{0} : \mu = 30
$$

versus the hypothesis

$$
H_{a} : \mu \gt 30
$$

where $\mu$ is the population mean RDI. Clearly, somehow we must figure out a way to decide between these hypotheses using the observed data, particularly the sample mean.

Before we go through the specifics, let’s set up the central ideas.

## Types of errors in Hypothesis Testing

The alternative hypotheses are typically of the form of the true mean being $\lt$, $\gt$ or $\neq$ to the hypothesized mean, such as $H_{a} : \mu \gt 30$ from our example. The null typically sharply specifies the mean, such as $H_{0} : \mu = 30$ in our example. More complex null hypotheses are possible, but are typically covered in later courses.

Note that there are four possible outcomes of our statistical decision process:

| Truth | Decide | Result |
|:-----|:--------|:------|
|$H_{0}$ | $H_{0}$ | Correctly accepting null | 
|$H_{0}$ | $H_{a}$ | Type I error | 
|$H_{a}$ | $H_{a}$ | Correctly reject null | 
|$H_{a}$ | $H_{0}$ | Type II Error | 

We will perform hypothesis testing by forcing the probability of a Type I error to be small. This approach consequences, which we can discuss with an analogy to court cases.

### Discussion relative to court cases

Consider a court of law and a criminal case. The null hypothesis is that the defendant is innocent. The rules requires a standard on the available evidence to reject the null hypothesis (and the jury to convict). The standard is specified loosely in this case, such as convict if the defendant appears guilty "Beyond reasonable doubt". In statistics, we can be mathematically specific about our standard of evidence.

Note the consequences of setting a standard. If we set a low standard, say convicting only if there circumstantial or greater evidence, then we would increase the percentage of innocent people convicted (type I errors). However, we would also increase the percentage of guilty people convicted (correctly rejecting the null).

If we set a high standard, say the standard of convicting if the jury has "No doubts whatsoever", then we increase the the percentage of innocent people let free (correctly accepting the null) while we would also increase the percentage of guilty people let free (type II errors).

## Building up a standard of evidence

Consider our sleep example again. A reasonable strategy would reject the null hypothesis if the sample mean, $\bar{X}$, was larger than some constant, say $C$. Typically, $C$ is chosen so that the probability of a Type I error, labeled $\alpha$, is 0.05 (or some other relevant constant) To reiterate, $\alpha$ = Type I error rate = Probability of rejecting the null hypothesis when, in fact, the null hypothesis is correct.

Let’s see if we can figure out what $C$ has to be. The standard error of the mean is $10/\sqrt{100} = 1$. Furthermore, under $H_{0}$ we know that $\bar{X} \sim \mathcal{N}(30, 1)$ (at least approximately) via the CLT. We want to chose C so that:

$$
P(\bar{X} \gt C; H_{0}) = 0.05
$$

The $95^{th}$ percentile of a normal distribution is $1.645$ standard deviations from the mean. So, if $C$ is set $1.645$ standard deviations from the mean, we should be set since the probability of getting a sample mean that large is only $5\%$. The $95^{th}$ percentile from a $\mathcal{N}(30, 1)$ is:

$$
C = 30 + 1 \times 1.645 = 31.645.
$$

So the rule "Reject $H_{0}$ when $\bar{X} \geq 31.645$" has the property that the probability of rejection is $5\%$ when $H_{0}$ is true.

In general, however, we don’t convert $C$ back to the original scale. Instead, we calculate how many standard errors the observed mean is from the hypothesized mean:

$$
Z = \frac{\bar{X} - \mu_{0}}{s/ \sqrt{n}}
$$

This is called a ***$Z$-score***. We can compare this statistic to standard normal quantiles. To reiterate, the $Z$-score is how many standard errors the sample mean is above the hypothesized mean. In our example:

$$
\frac{32-30}{10/\sqrt{100}} = 2
$$

Since $2$ is greater than $1.645$ we would reject. Setting the rule "We reject if our $Z$-score is larger than 1.645" controls the Type I error rate at $5\%$. We could write out a general rule for this alternative hypothesis as reject whenever $\sqrt{n}(\bar{X} − \mu_{0})/s > Z_{1−\alpha}$ where $\alpha$ is the desired Type I error rate. Because the Type I error rate was controlled to be small, if we reject we know that one of the following occurred:

1. the null hypothesis is false,
2. we observed an unlikely event in support of the alternative even though the null is true,
3. our modeling assumptions are wrong.

The third option can be difficult to check and at some level all bets are off depending on how wrong we are about our basic assumptions. So for this course, we speak of our conclusions under the assumption that our modeling choices (such as the iid sampling model) are correct, but do so wide eyed acknowledging the limitations of our approach.

## General Rules

We developed our test for one possible alternatives. Here’s some general rules for the three most important alternatives.

Consider the $Z$ test for $H_{0} : \mu = \mu_{0}$ versus: $H_{1} : \mu \lt \mu_{0}$, $H_{2} : \mu \neq \mu_{0}$, $H_{3} : \mu \gt \mu_{0}$. Our test statistic

$$
TS = \frac{\bar{X} - \mu_{0}}{S/\sqrt{n}}
$$

We reject the null hypothesis when:

- $H_{1}: TS \leq Z_{\alpha} = -Z_{1-\alpha}$
- $H_{2}: |TS| \geq Z_{1-\alpha/2}$
- $H_{3}: TS \geq Z_{1-\alpha/2}$

respectively.

## Summary notes

- We have fixed $\alpha$ to be low, so if we reject $H_{0}$ (either our model is wrong) or there is a low probability that we have made an error.
- We have not fixed the probability of a type II error, $\beta$; therefore we tend to say "Fail to reject H0" rather than accepting $H_{0}$.
- Statistical significance is no the same as scientific significance.
- The region of $TS$ values for which you reject $H_{0}$ is called the rejection region.
- The $Z$ test requires the assumptions of the CLT and for $n$ to be large enough for it to apply.
- If $n$ is small, then a Gosset’s $t$ test is performed exactly in the same way, with the normal quantiles replaced by the appropriate Student’s $t$ quantiles and $n − 1$ df.
- The probability of rejecting the null hypothesis when it is false is called power
- Power is a used a lot to calculate sample sizes for experiments.

## Example reconsidered

Consider our example again. Suppose that $n = 16$ (rather than $100$). The statistic

$$
\frac{\bar{X} - 30}{s/\sqrt{16}}
$$

follows a $t$ distribution with $15$ degrees of freedom under $H_{0}$.

Under $H_{0}$, the probability that it is larger that the $95^{th}$ percentile of the $t$ distribution is $5\%$. The $95^{th}%$ percentile of the $T$ distribution with 15 degrees of freedom is $1.7531$ (obtained via Python `scipy.stats.t.ppf(.95, 15)`). Assuming that everything but the sample size is the same, our test statistic is now $\sqrt{16} (32-30)/10 = 0.8$. Since $0.8$ is not larger than $1.75$, we now fail to reject.

## Two sided tests

In many settings, we would like to reject if the true mean is *different* than the hypothesized, not just larger or smaller. I other words, we would reject the null hypothesis if in fact the sample mean was much larger or smaller than the hypothesized mean. In our example, we want to test the alternative $H_{a} : \mu \neq 30$. We will reject if the test statistic, $0.8$, is either too large or too small. Then we want the probability of rejecting under the null to be $5\%$, split equally as $2.5\%$ in the upper tail and $2.5\%$ in the lower tail. 

Thus we reject if our test statistic is larger than `t.ppf(.975, 15)` or smaller than `t.ppf(.025, 15)`. This is the same as saying: reject if the absolute value of our statistic is larger than `t.ppf(0.975, 15)` = 2.1314. In this case, since our test statistic is $0.8$, which is smaller than $2.1314$, we fail to reject the two sided test (as well as the one sided test).

If you fail to reject the one sided test, then you would fail to reject the two sided test. Because of its larger rejection region, two sided tests are the norm (even in settings where a one sided test makes more sense).

## T test in Python

Let’s try the $t$ test on the pairs of fathers and sons in Galton’s data.

In [2]:
father_son = pd.read_csv('./data/father_son.csv')

In [3]:
pttest(father_son['sheight'] - father_son['fheight'], conf_int=0.95)


	 One Sample t-test

data: 	father_son['sheight'] - father_son['fheight'], conf_int=0.95
t = 11.789, df = 1077, p-value < 2.957226369228171e-30
0.95 percent confidence interval:
0.831 1.163
sample esimates:
mean of x
0.997


## Connections with confidence intervals

Consider testing $H_{0} : \mu = \mu_{0}$ versus $H_{a} : \mu \neq \mu_{0}$. Take the set of all possible values for which you fail to reject $H_{0}$, this set is a $(1-\alpha)100 \%$ confidence interval for $\mu$.

The same works in reverse; if a $(1-\alpha)100 \%$ interval contains $\mu_{0}$, then we *fail to* reject $H_{0}$.

In other words, two sided tests and confidence intervals agree. 

Consider the previous example. We were interested in whether or not, father's and son heights were equal. We got the answer that, no, they weren't as far as the hypnosis test was concerned. We also got the answer that the interval did not contain zero. You might be wondering whether or not those two statements could ever disagree and it turns out that no they can't. Checking whether or not munot is in the interval is identical to performing the two
sided hypothesis test with the caveat that the alpha that you use for
the interval, for the hypothesis test. Has to be equal to 1 minus alpha for
the confidence interval. In other words if you
construct a 95% interval, and just look for whether or
not munot is in that interval and failed to reject if it's in that interval
and reject if it's outside the interval. That is the same as performing that
hypotheses test with an alpha level, with a type point error of exactly alpha. This is stated here in
this slide by saying the confidence interval can be constructed
as the set of all possible values for which you fail to reject
a null hypothesis. 

## Two group intervals

Doing group tests is now straightforward given that we’ve already covered independent group T intervals. Our rejection rules are the same, the only change is how the statistic is calculated. However, the form is familiar:

$$
\frac{\text{Estimate} - \text{Hypothesized Value}}{\text{Standard Error}}
$$

Consider now testing $H_{0} : \mu_{1} = \mu_{2}$. Our statistic is:

$$
\frac{\bar{X}_{1} - \bar{X}_{2} - (\mu_{1} - \mu_{0})}{S_{p} \sqrt{\frac{1}{n_{1}} + \frac{1}{n_{2}}}}
$$

For the equal variance case and

$$
\frac{\bar{X}_{1} - \bar{X}_{2} - (\mu_{1} - \mu_{0})}{ \sqrt{\frac{S_{1}^{2}}{n_{1}} + \frac{S_{2}^{2}}{n_{2}}}}
$$

For unequal variance.

Let's just go through an example.

## Example `chickWeight` data

Recall that we reformatted this data as follow:

In [4]:
import numpy as np
import pandas as pd
from scipy.stats import ttest_ind

data = pd.read_csv('./data/wideCW.csv')

wideCW14 = data[data['Diet'].isin([1,4])]
d1 = wideCW14[wideCW14['Diet'] == 1]['gain']
d4 = wideCW14[wideCW14['Diet'] == 4]['gain']

def pindipendent_test(g1, g2, var_equal = True, confidence_level = 0.95):
    """
    Function computing Two Sample T test
    s1: np.array for the first sample
    s2: np.array for the second sample
    """
    print("\n\t Two Sample t-test\n")
    k = ttest_ind(g1.dropna(), g2.dropna())
    print("t = {}, df = {}, p-value = {}".format(np.round(k.statistic, 3), np.round(k.df, 4), np.round(k.pvalue, 5)))
    print("alternative hypothesis: true difference in means is not equal to 0")
    print("{} percent confidence interval:".format(confidence_level))
    print("{} {}".format(np.round(k.confidence_interval(confidence_level=confidence_level).low,2), np.round(k.confidence_interval(confidence_level=confidence_level).high,2)))
    print("sample estimates:")
    print("mean in group 1 mean in group 2")
    print("\t {} \t {}".format(np.round(g1.mean(),2), np.round(g2.mean(),2)))

pindipendent_test(d1, d4)


	 Two Sample t-test

t = -2.725, df = 23.0, p-value = 0.01207
alternative hypothesis: true difference in means is not equal to 0
0.95 percent confidence interval:
-108.15 -14.81
sample estimates:
mean in group 1 mean in group 2
	 136.19 	 197.67


## Exact binomial test

Recall this problem. Suppose a friend ha $8$ childrean, $7$ of which are girls and none are twins. Perform the relevan hypothesis test. 

$H_{0} : p = 0.5 \text{ versus } H_{a} : p \gt 0.5$

What is the relevant rejection region so that the probability of rejecting is (less than) $5 \%$?

In [5]:
from scipy.stats import binom

print("Rejection region \t Type I error rate")
for i in np.arange(8,-1,-1):
    print("[{}:8]           \t {}".format( 8-i, binom.cdf(i,8, p =0.5)))

Rejection region 	 Type I error rate
[0:8]           	 1.0
[1:8]           	 0.99609375
[2:8]           	 0.96484375
[3:8]           	 0.85546875
[4:8]           	 0.63671875
[5:8]           	 0.36328125
[6:8]           	 0.14453125
[7:8]           	 0.03515625
[8:8]           	 0.00390625


Thus if we reject under 7 or 8 girls, we will have a less than $5 \%$ chance of rejecting under the null hypothesis.

It’s impossible to get an exact 5% level test for this case due to the discreteness of the binomial. The closest is the rejection region [7 : 8]. Further note that an alpha level lower than 0.0039 is not attainable. For larger sample sizes, we could do a normal approximation. 

Extended this test to two sided test isn’t obvious. Given a way to do two sided tests, we could take the set of values of $p_{0}$ for which we fail to reject to get an exact binomial confidence interval (called the *Clopper/Pearson interval*, by the way). We’ll cover two sided versions of this test when we cover P-values.